cap-0018 - Implementation after removing low-level auth checks starting from v10

sisuresh opened the pull request:
# Description

Implements https://github.com/stellar/stellar-protocol/blob/master/core/cap-0018.md


# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

jonjove commented on src/transactions/AllowTrustOpFrame.cpp:
```diff
@@ -136,6 +137,13 @@ AllowTrustOpFrame::doCheckValid(uint32_t ledgerVersion)
         return false;
     }
+
+    if (!trustLineFlagIsValid(mAllowTrust.authorize))
```
Transactions subsystem shouldn't have dependencies in the invariants subsystem. The transaction subsystem is protocol-level, whereas the invariants subsystem isn't.

sisuresh replied:
This method should be shared between the transaction and invariants subsystem. What do you think about renaming `InvariantUtils.h/cpp` and moving it to `src/util`?

jonjove commented on src/invariant/InvariantUtils.cpp:
```diff
@@ -0,0 +1,14 @@
+bool
+trustLineFlagIsValid(uint32_t flag)
+{
+    return flag <= AUTHORIZED_FLAG;
```
This should probably use bitwise comparisons rather than inequality comparisons, to reflect the fact that this is a flags field.

jonjove commented on src/invariant/LedgerEntryIsValid.cpp:
```diff
@@ -174,7 +175,8 @@ LedgerEntryIsValid::checkIsValid(TrustLineEntry const& tl, uint32 version) const
     }
-    if ((tl.flags & ~MASK_TRUSTLINE_FLAGS) != 0)
+    if ((tl.flags & ~MASK_TRUSTLINE_FLAGS) != 0 ||
+        !trustLineFlagIsValid(tl.flags))
```
I think `trustLineFlagIsValid` should encompass all validity checks on the flags field, including no extraneous bits set.

jonjove commented on src/transactions/OfferExchange.cpp:
```diff
@@ -50,7 +50,7 @@ canSellAtMost(LedgerTxnHeader const& header, LedgerTxnEntry const& account,
     }
 
-    if (trustLine && trustLine.isAuthorized())
+    if (trustLine && trustLine.isAuthorizedToMaintainLiabilities())
```
Do we actually need to check authorization here at all starting from version 10? Same thing for the other overload.

sisuresh replied:
This is a good point. We don't need to check as the authorization is checked prior to this for v10 and up. I did this because it looked cleaner than `header.current().ledgerVersion >= 10 || trustLine.isAuthorized()`, but I think the intention is clearer if we only check the authorization pre v10. I'll make the change.

MonsieurNicolas commented on src/invariant/LiabilitiesMatchOffers.cpp:
```diff
@@ -96,6 +98,28 @@ checkAuthorized(std::shared_ptr<LedgerEntry const> const& current)
+            if (sellingLiabilitiesInc || buyingLiabilitiesInc)
+            {
+                return fmt::format(
+                    "Liabilities increased on unauthorized trust line {}",
```
error message is confusing: this can only be reached if `AUTHORIZED_TO_MAINTAIN_LIABILITIES_FLAG` is set.

The code would be clearer if it was written like this:

```
if (!isAuthorized())
{
     if (isAuthorizedToMaintainLiabilities())
     {
         // check that liabilities can only decrease
     }
     else
     {
          // check that there cannot be any liabilities
     }
}
```

sisuresh replied:
done

MonsieurNicolas commented on src/transactions/OfferExchange.cpp:
```diff
@@ -50,7 +50,8 @@ canSellAtMost(LedgerTxnHeader const& header, LedgerTxnEntry const& account,
 
-    if (trustLine && trustLine.isAuthorized())
+    if (trustLine &&
+        (header.current().ledgerVersion >= 10 || trustLine.isAuthorized()))
```
Is this true because in protocol 10 and above we delete offers when trust is revoked? If so, a comment would be welcome.

MonsieurNicolas replied:
now that I think of it, I am not sure this buys us much vs the defensive coding version `trustLine.isAuthorizedToMaintainLiabilities()`

sisuresh replied:
switched to `isAuthorizedToMaintainLiabilities`

MonsieurNicolas commented on src/transactions/TransactionUtils.cpp:
```diff
@@ -230,7 +231,8 @@ addBalance(LedgerTxnHeader const& header, LedgerTxnEntry& entry, int64_t delta)
         }
-        if (!isAuthorized(entry))
+
+        if (header.current().ledgerVersion < 10 && !isAuthorized(entry))
```
I understand that this "cannot be reached", but as this is super critical code, I'd rather keep it as `!isAuthorizedToMaintainLiabilities(entry)` (defensive)

sisuresh replied:
done

MonsieurNicolas commented on src/transactions/TransactionUtils.cpp:
```diff
@@ -293,12 +295,12 @@ addBuyingLiabilities(LedgerTxnHeader const& header, LedgerTxnEntry& entry,
     {
-        auto& tl = entry.current().data.trustLine();
-        if (!isAuthorized(entry))
+        if (header.current().ledgerVersion < 10 && !isAuthorized(entry))
```
same thing here. I'd prefer to just check for `isAuthorizedToMaintainLiabilities` (defensive coding)

sisuresh replied:
done

MonsieurNicolas commented on src/transactions/TransactionUtils.cpp:
```diff
@@ -497,15 +499,15 @@ getMaxAmountReceive(LedgerTxnHeader const& header, LedgerEntry const& le)
     {
-        int64_t amount = 0;
-        if (isAuthorized(le))
+        auto const& tl = le.data.trustLine();
```
again, maybe just stylistic, but using `isAuthorizedToMaintainLiabilities` keeps the code easy to understand

sisuresh replied:
I'm not sure if it makes much sense here. That line can only be hit if the ledger version check above faile, where the `AUTHORIZED_TO_MAINTAIN_LIABILITIES_FLAG` flag shouldn't even be set.

jonjove commented on src/util/ValidityUtils.cpp:
```diff
@@ -2,6 +2,7 @@
 // under the Apache License, Version 2.0. See the COPYING file at the root
 // of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
 
+#include "util/ValidityUtils.h"
```
I think it probably makes the most sense to just incorporate this into TransactionUtils.{h,cpp}. Sorry I missed responding to your earlier comment about this.

jonjove commented on src/transactions/TransactionUtils.cpp:
```diff
@@ -232,7 +232,7 @@ addBalance(LedgerTxnHeader const& header, LedgerTxnEntry& entry, int64_t delta)
         }
 
-        if (header.current().ledgerVersion < 10 && !isAuthorized(entry))
+        if (!isAuthorizedToMaintainLiabilities(entry))
```
This doesn't make any sense to me. I see the discussion https://github.com/stellar/stellar-core/pull/2387#discussion_r390052332 (GitHub is giving me a hard time about replying to it....). It should be `!isAuthorized`, because you can't add balance if you are only authorized to maintain liabilities. Same remark applies for this change in `addBuyingLiabilities` and `addSellingLiabilities`: you can't add liabilities if you are only authorized to maintain liabilities. Am I missing something?

sisuresh replied:
The balance can change if an existing offer was filled. Similar thing with the add liabilities methods - liabilities can be removed when an offer crosses. The offer can also be deleted.

MonsieurNicolas commented on src/transactions/TransactionUtils.cpp:
```diff
@@ -497,15 +498,15 @@ getMaxAmountReceive(LedgerTxnHeader const& header, LedgerEntry const& le)
-        if (isAuthorized(le))
+        auto const& tl = le.data.trustLine();
+        int64_t amount = tl.limit - tl.balance;
+        if (header.current().ledgerVersion >= 10)
```
I am not sure I understand the behavior for protocol 10+:
if not authorized, shouldn't we always return 0? Right now it's going to return whatever `tl.limit` is.

Or is it a situation where "we know" that the caller will never call this function if `isAuthorizedToMaintainLiabilities()` is `false`? Which is at best error prone.

If so, I'd rather keep the old code (just use `isAuthorizedToMaintainLiabilities` instead of `isAuthorized`)

sisuresh replied:
Yeah this is assuming the authorization was already checked. I'll change this back to the old code and use `isAuthorizedToMaintainLiabilities`.

sisuresh replied:
Yeah this is assuming the authorization was already checked. I'll change this back to the old code and use `isAuthorizedToMaintainLiabilities`.

jonjove commented on src/transactions/test/OfferTests.cpp:
```diff
@@ -674,350 +674,435 @@ TEST_CASE("create offer", "[tx][offers]")
-                for_versions_from(10, *app, [&] {
-                    issuer.pay(b1, usd, 20000);
+                        // offer is sell 150 USD for 100 USD; sell USD @ 1.5 /
+                        // buy IRD @ 0.66
```
Nitpick: typo "IRD" -> "IDR" (this typo happens several times in this file)

sisuresh replied:
Fixed

jonjove commented on src/transactions/test/OfferTests.cpp:
```diff
@@ -674,350 +674,435 @@ TEST_CASE("create offer", "[tx][offers]")
+                SELL_AUTHORIZED_TO_MAINTAIN_LIABILITIES
+            };
+            auto multipleOffers = [&](AuthState authState) {
+                auto a1CheckSetAuthToMaintainLiabilities = [&]() {
```
Nitpick: same thing as other "check" function. Do you mean something like "a1MaybeSet..."?

sisuresh replied:
Fixed. I think I meant "checkToSet..." but "maybe" sounds better.

jonjove commented on src/transactions/test/OfferTests.cpp:
```diff
@@ -674,350 +674,435 @@ TEST_CASE("create offer", "[tx][offers]")
+                SECTION("offer crosses own")
+                {
+                    for_all_versions(*app, [&] {
+                        issuer.pay(a1, usd, 20000);
```
Missing `a1CheckSet...`?

sisuresh replied:
This tests that an account can't add an offer to cross with one of its own existing offers. If we're authorized to maintain liabilities, then the add offer will fail for a different reason (`*_NOT_AUTHORIZED`). We already test that an offer can't be added when an account is in this state in `AllowTrustTests`.

sisuresh commented on src/transactions/test/PathPaymentTests.cpp:
```diff
@@ -1419,214 +1451,258 @@ TEST_CASE("pathpayment", "[tx][pathpayment]")
-        auto mm23 = root.create("mm23", minBalance3);
-        auto mm34 = root.create("mm34", minBalance3);
+    auto usesWholeBestOffer = [&](bool testAuthorizedToMaintainLiabilities) {
+        auto checkSetAuthToMaintainLiabilities =
```
Fixed

sisuresh commented:
> This looks good. I left a few very minor comments on one of the tests. But I also had some thoughts while reviewing the PR that aren't necessarily specific to any single line:
> 
> * I'm not sure that the defensive programming approach that we took in this PR is the right one. I am of the opinion that, starting from protocol 10 (or maybe 13) we should throw if we reach `addBalance`, `addBuyingLiabilities`, etc with `!isAuthorizedToMaintainLiabilities`. I think this is the correct approach because we are already depending on the client code to check that it is correct between `isAuthorizedToMaintainLiabilities` and `isAuthorized`. If the client code has permitted those functions to be reached in the case mentioned above, then there is definitely a logic error. We shouldn't carry on processing the transaction if there is definitely a logic error, instead we should return `txINTERNAL_ERROR`.

I added a commit that throws if we reach those methods starting in protocol 10. I like this approach as the assumptions made are clear and enforced.

> * We added defensive programming checks to `getMaxAmountReceive` but they are missing from the analogous function `getAvailableBalance` (which didn't require any changes in this PR).

I added an authorization check in `getAvailableBalance`. I think this makes sense because there's no reason to retrieve the balance of an unauthorized asset (since nothing can be done with that asset).

> * The changes in the first commit were based off of the assumption that we always check authorization before reaching those functions, but that isn't true on the path through `releaseLiabilities` (see, for example, ManageOfferOpFrameBase.cpp:224 and note that the offer loaded there does not necessarily have the same assets as specified in the operation). Now assuming that everything works correctly, this shouldn't cause any issues because an offer in the ledger should have authorized trust lines. But we took similar precautions at OfferExchange.cpp:1112.

This is an interesting scenario. The authorization shouldn't matter since `releaseLiabilities` is removing offers. However, if there are unauthorized orders in the ledger, then we will throw in `addBuyingLiabilities` and `addSellingLiabilities` (with my latest change). Because of the new invariants, I think the check at `OfferExchange.cpp:1112` is unnecessary because `releaseLiabilities` is called on the existing offer, which will do the auth check anyways. If you agree, I will remove that check.

sisuresh commented on src/transactions/TransactionUtils.cpp:
```diff
@@ -429,14 +448,18 @@ getAvailableBalance(LedgerTxnHeader const& header, LedgerEntry const& le)
     }
     else if (le.data.type() == TRUSTLINE)
     {
+        if (ledgerVersionMinV10 && !isAuthorizedToMaintainLiabilities(le))
```
I just thought it was easier to read this way. But using `checkAuthorization` will work as well, and will allow us to keep the throw logic in one spot. I'll make the change.

sisuresh commented:
> > This is an interesting scenario. The authorization shouldn't matter since releaseLiabilities is removing offers. However, if there are unauthorized offers in the ledger, we will throw in addBuyingLiabilities and addSellingLiabilities (with my latest change). Because of the new invariants, I think the check at OfferExchange.cpp:1112 is unnecessary because releaseLiabilities is called on the existing offer at OfferExchange.cpp:1100, which will do the auth check anyways. If you agree, I will remove the check on OfferExchange.cpp:1112.
> 
> Yes, I agree that the check on OfferExchange.cpp:1112 is unnecessary now. I think it would be worth adding a comment where `releaseLiabilitites` is used on OfferExchange.cpp:1099 indicating that this is being checked there. Have you checked that everything makes sense at other places where `releaseLiabilities` is used, such as AllowTrustOpFrame.cpp:135 (there are a few other places too)?

Yeah the other `releaseLiabilities` calls look good. Related to this - Do you think it makes sense to throw if unauthorized liabilities are being released? Obviously the scenario shouldn't happen, but if it was encountered for some reason, you could make the argument that the offers should still be pulled (and log an error).

MonsieurNicolas commented:
r+ 1ca2deb5e30f35f7d2a2917e70bfd6c9cf9df6b8
